### DISCUSSION TITLE
How to understand which path FontForge considers to be the inner vs outer boundary?

### DISCUSSION THREAD

**Discussion by rybak** [fontforge/fontforge#5443] [Q&A] [answered]
_Disclaimer: I've only started using FontForge two hours ago. I'm sorry for the wall of text. I'm just trying to write down all of my understanding, because there's obviously a gap in it somewhere._

### Tutorial

<details><summary>Additional context about me reading the tutorial (click to expand)</summary>

I am trying to understand how holes in glyphs are understood by FontForge. I am reading the tutorial section: [4.2. Creating the letter “o” – consistent directions](https://fontforge.org/docs/tutorial/editexample2.html#creating-the-letter-o-consistent-directions). I skipped ahead – it is possible that the answer to my question was in previous sections. The tutorial says:

> ### Stages in editing “o”
>
> Notice that the first outline is drawn clockwise and the second counter-clockwise. This change in drawing direction is important.

I understand this notion intuitively from my very basic knowledge of [winding numbers in topology](https://en.wikipedia.org/wiki/Winding_number). I also  see a tiny band of shading near the paths, which _seems_ to suggests where the "filling" of the glyph is.
</details>

### The problem

Despite the above, I am unable to overcome the "hole which should be empty is filled-in" problem, as the passage in the tutorial explains:

> If you fail to alternate directions between outer and inner boundaries you may get results like the one on the left ![image](https://github.com/fontforge/fontforge/assets/624072/261fdb08-75ab-4c15-b0c7-a208d2f17bdb)
>
> [...]
> 
> The command [Element->Correct Direction](https://fontforge.org/docs/ui/menus/elementmenu.html#elementmenu-correct) will look at each selected contour, figure out whether it qualifies as an outer or inner contour and will reverse the drawing direction when the contour is drawn incorrectly.

The menu "Element->Correct Direction" is disabled for me when I select the inner boundary.

For some reason, the digit six is OK in the preview of the Outline Glyph view, but broken in Font View:
![image](https://github.com/fontforge/fontforge/assets/624072/08a21689-47b0-4ff5-8af8-ecae96a3c104)

#### Find Problems

When I do "Element > Find Problems > Paths > Check outermost paths clockwise", I get: "This glyph self-intersects", but [the section "Finding common font problems automagically § Intersecting Paths"](https://fontforge.org/docs/ui/dialogs/problems.html#problems-intersecting) doesn't explain how to resolve it, so I am stuck.

![image](https://github.com/fontforge/fontforge/assets/624072/6b967f5b-41a1-40e7-b0d8-77152c23aaf8)

### Order of points

I know that when I press <kbd>Tab</kbd> on my keyboard repeatedly, I iterate over the points in their order. Here's what seems to be a problem to me lies: for some reason this iteration jumps from "outer boundary" to "inner boundary". It seems that FontForge considers this to be a singular path, rather than two paths (inner and outer).

Here's a screen recording demonstrating my confusion. I start near the center of digit six, iterate over the outer boundary, but after going round what should be the outer boundary it jumps to the inner boundary – I lose track of where the selection goes at ~16 seconds. Is this expected?

https://github.com/fontforge/fontforge/assets/624072/e5d8295b-3e0c-406f-8b35-5fbe67bebb33

Is it possible to make FontForge split the singular path into two paths?

_In case it's relevant: I am not making this glyph from scratch. I am taking bits and pieces from other glyphs and haphazardly joining them together._

**Comment by rybak** [comment]
> For some reason, the digit six is OK in the preview of the Outline Glyph view, but broken in Font View:

I have managed to fix this problem. I found that one of the points in the inner boundary was actually two points, which weren't connected. After connecting them:

- the Font View is now OK
- "Find Problems" says "No problems found"

Is there any way of finding such gaps faster?

**Comment by skef** [comment, marked as answer]
"Open" contours have a red-ish shadow, while closed contours have a grey shadow. 

Also, an open contour has to be open at its starting point, and the starting point is depicted with an arrow. So If you have a contour with a red-ish shadow that seems to be closed there is going to be some problem with the starting point -- probably an ending point right next to it.

**Reply by rybak** [reply]
> "Open" contours have a red-ish shadow, while closed contours have a grey shadow.

Yep, I figured this out by trial and error eventually.

> an open contour has to be open at its starting point

This is what I was missing, thanks. I increased the size of the handles in preferences, but I couldn't find a way to customize the size of the arrow. It's way too small on my screen.
